### PR TITLE
lua: Upgrade the LuaJIT version to 2022-12-22

### LINKED ISSUE #25763: Upgrade the LuaJIT version used in Envoy
*Title*: *Upgrade the LuaJIT version used in Envoy*

*Description*:
The current LuaJIT version used in Envoy is 1d8b747c161db457e032a023ebbff511f5de5ec2, which is created on 2020-10-12, more than 2 years ago. Plenty of bugfix and new features happen after that. 

For example, https://github.com/envoyproxy/envoy/blob/6c18a51ea75b8b184e65acdb8364403df4b35e37/tools/code_format/check_format.py#L726-L730 claims that lua_pushlightuserdata is unsafe but this problem is already fixed in the latest LuaJIT.

I recommend upgrading the version to https://github.com/LuaJIT/LuaJIT/commit/a04480e311f93d3ceb2f92549cad3fffa38250ef, which is also the version used in the latest OpenResty: https://github.com/openresty/luajit2/commits/v2.1-20230119.

If this is acceptable, I can work on this and update the [patch](https://github.com/envoyproxy/envoy/blob/main/bazel/foreign_cc/luajit.patch) we apply to the LuaJIT.

### PR DESCRIPTION
**Pull request by spacewander**
Commit Message: lua: Upgrade the LuaJIT version to 2022-12-22
Additional Description: The current LuaJIT version used in Envoy is 1d8b747c161db457e032a023ebbff511f5de5ec2, which is created on 2020-10-12, more than 2 years ago. Plenty of bugfix and new features happen after that.
Risk Level: Middle
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
Fixes #25763

### CODE REVIEW

File: bazel/repository_locations.bzl [hunk 1]
```diff
@@ -425,8 +425,8 @@ REPOSITORY_LOCATIONS_SPEC = dict(
         project_url = "https://luajit.org",
         # The last release version, 2.1.0-beta3 has a number of CVEs filed
         # against it. These may not impact correct non-malicious Lua code, but for prudence we bump.
```
Comment (phlax) [inline bazel/repository_locations.bzl:427]: this comment seems out of date - perhaps we can change to something like "LuaJIT only provides rolling releases"

File: bazel/repository_locations.bzl [hunk 2]
```diff
@@ -425,8 +425,8 @@ REPOSITORY_LOCATIONS_SPEC = dict(
...
-        version = "1d8b747c161db457e032a023ebbff511f5de5ec2",
-        sha256 = "20a159c38a98ecdb6368e8d655343b6036622a29a1621da9dc303f7ed9bf37f3",
+        version = "a04480e311f93d3ceb2f92549cad3fffa38250ef",
```
Comment (phlax) [inline bazel/repository_locations.bzl:428]: i was going to comment that we should use a tagged release as new versions would be flagged by our deps system, but then i read https://github.com/LuaJIT/LuaJIT/issues/563#issuecomment-1206294291 so looks like that wont happen

**Review by phlax — COMMENTED** [review]
thanks for working on this @spacewander 

small nit inline

File: bazel/repository_locations.bzl (hunk 2 above)
Comment (spacewander) [reply bazel/repository_locations.bzl:427]: Yes, it seems that Mike Pall is tired of making new release.

**Review by phlax — APPROVED** [review]
lgtm, thanks @spacewander

**Pull request merged**